src/java.desktop/share/native/libawt/java2d/Trace.h

```diff
@@ -54,79 +54,13 @@ JNIEXPORT void JNICALL
 J2dTraceImpl(int level, jboolean cr, const char *string, ...);
 
 #ifndef DEBUG
-#define J2dTrace(level, string)
-#define J2dTrace1(level, string, arg1)
-#define J2dTrace2(level, string, arg1, arg2)
-#define J2dTrace3(level, string, arg1, arg2, arg3)
-#define J2dTrace4(level, string, arg1, arg2, arg3, arg4)
-#define J2dTrace5(level, string, arg1, arg2, arg3, arg4, arg5)
-#define J2dTrace6(level, string, arg1, arg2, arg3, arg4, arg5, arg6)
-#define J2dTrace7(level, string, arg1, arg2, arg3, arg4, arg5, arg6, arg7)
-#define J2dTrace8(level, string, arg1, arg2, arg3, arg4, arg5, arg6, arg7, arg8)
-#define J2dTraceLn(level, string)
-#define J2dTraceLn1(level, string, arg1)
-#define J2dTraceLn2(level, string, arg1, arg2)
-#define J2dTraceLn3(level, string, arg1, arg2, arg3)
-#define J2dTraceLn4(level, string, arg1, arg2, arg3, arg4)
-#define J2dTraceLn5(level, string, arg1, arg2, arg3, arg4, arg5)
-#define J2dTraceLn6(level, string, arg1, arg2, arg3, arg4, arg5, arg6)
-#define J2dTraceLn7(level, string, arg1, arg2, arg3, arg4, arg5, arg6, arg7)
-#define J2dTraceLn8(level, string, arg1, arg2, arg3, arg4, arg5, arg6, arg7, arg8)
+#define J2dTrace(level, ...)
+#define J2dTraceLn(level, ...)
```

Comment:
This looks fine, but maybe the string argument could be included as a parameter? Would that improve usability, like better error messages when no arguments are provided, or enhance auto-completion in the IDE?"

Reply:
There is a trailing comma problem when passing no extra args. I didn't figure out how to solve it without `__VA_OPT__` or compiler-specific tricks, so I moved the string into variadic part, so that it always has at least one parameter.

Reply:
I think that the benefits (having an arbitrary number of arguments) outweigh the lack of a check for the string argument type. So, I'd let's put some comments here about future use of __VA_OPT__ and integrate this change.

Reply:
I think we still have the same benefits, and the first argument has to be string… It wasn't enforced at the macro level, but it was suggested by the name of the argument.

https://github.com/openjdk/jdk/blob/867aee60dd12c4ea232aaf03f14a45a3d0f91851/src/java.desktop/share/native/libawt/java2d/Trace.c#L38C38-L38C44

The functions that implement traces, `J2dTraceImpl`, still require that the first argument is a string, therefore we should get a compilation error if the first parameter isn't a string.